File: Java/NQueensII.java
class Solution {
    private int res = 0;

    private Set<Integer> cols = new HashSet<>();
    private Set<Integer> diff = new HashSet<>();
    private Set<Integer> sum = new HashSet<>();


    public int totalNQueens(int n) {
        if(n <= 0) return res;
        dfs(0, n);
        return res;
    }

    private void dfs(int row, int n) {
        if(row == n) {
            res++;
            return;
        }
        for(int col = 0; col < n; col++) {
            if(!cols.contains(col) && !diff.contains(col - row) && !sum.contains(col + row)) {
                cols.add(col); diff.add(col - row); sum.add(col + row);
                dfs(row + 1, n);
                cols.remove(col); diff.remove(col - row); sum.remove(col + row);
            }
        }
    }
}

File: Go/nqueens_ii.go
func totalNQueens(n int) int {
    res := 0
    if  n <= 0 {return res}

    cols := make(map[int]int, n)
    diff := make(map[int]int, n)
    sum := make(map[int]int, n)

    var dfs func(row int)
    dfs = func(row int) {
        if row == n {
            res++;
            return;
        }
        for col := 0; col < n; col++ {
            if cols[col] == 0 && diff[col - row] == 0 && sum[col + row] == 0 {
                cols[col] = 1
                diff[col - row] = 1
                sum[col + row] = 1
                dfs(row + 1)
                cols[col] = 0
                diff[col - row] = 0
                sum[col + row] = 0
            }
        }
    }
    
    dfs(0)
    return res
}